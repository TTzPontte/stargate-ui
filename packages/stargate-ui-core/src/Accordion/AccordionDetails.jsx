import React from 'react';
import { useStyles } from '@pontte/stargate-ui-styles';
import clsx from 'clsx';
import PropTypes from 'prop-types';

import Factory from '../Factory';
import AccordionContext from './AccordionContext';

const styles = () => {
  const details = {
    '&:not($detailsExpanded)': {
      height: 0,
      overflow: 'hidden',
    },
  };

  const detailsExpanded = {
    height: 'auto',
  };

  return { details, detailsExpanded };
};

const AccordionDetails = React.forwardRef((props, ref) => {
  const {
    children,
    className: inheritedClassName,
    ...factoryProps
  } = props;

  const { opened } = React.useContext(AccordionContext);

  const [
    {
      details,
      detailsExpanded,
    }
  ] = useStyles(styles);
  const classDetails = clsx(
    details,
    {
      [detailsExpanded]: opened,
    },
    inheritedClassName,
  );

  return (
    <Factory
      ref={ref}
      className={classDetails}
      {...factoryProps}
    >
      <Factory
        paddingX={2}
        paddingTop={1}
        paddingBottom={2}
        children={children}
      />
    </Factory>
  );
});

AccordionDetails.displayName = 'AccordionDetails';

AccordionDetails.propTypes = {
  /**
   * Add DOM element. Can be a React component or HTML element.
   */
  children: PropTypes.node.isRequired,
  /**
   * Add a new CSS class to `className` property.
   * @default undefined
   */
  className: PropTypes.string,
};

export default AccordionDetails;
