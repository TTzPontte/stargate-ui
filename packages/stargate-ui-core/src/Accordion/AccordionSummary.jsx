import React from 'react';
import { useStyles } from '@pontte/stargate-ui-styles';
import clsx from 'clsx';
import PropTypes from 'prop-types';

import Factory from '../Factory';
import Typography from '../Typography';
import Grid from '../Grid';
import AccordionContext from './AccordionContext';
/**
 * @todo change location to @pontte/stargate-ui-icons when has been fixed
 */
import { ArrowFlatRight } from '../icons';

const styles = (theme) => {
  const { resets, transition } = theme;

  const summary = {
    ...resets.button,
    width: '100%',
  };

  const summaryIcon = {
    transition: transition('transform').ease,
    transform: ({ opened }) => (
      opened ? 'rotate(-90deg)' : 'rotate(0)'
    ),
  };

  return {
    summary,
    summaryIcon,
  };
};

const AccordionSummary = React.forwardRef((props, ref) => {
  const {
    children,
    className: inheritedClassName,
    color = 'primary',
    ...typographyProps
  } = props;

  const { opened, onToggle } = React.useContext(AccordionContext);

  const [{ summary, summaryIcon: classSummaryIcon }] = useStyles(styles, { opened });
  const classSummary = clsx(summary, inheritedClassName);

  return (
    <Factory
      ref={ref}
      paddingX={2}
      paddingY={2}
      {...typographyProps}
      className={classSummary}
      type="button"
      element="button"
      onClick={onToggle}
    >
      <Grid
        row
        alignItems="center"
        spacing={2}
      >
        <Grid col xs>
          <Typography
            weight="bold"
            color={color}
            gutter={0}
            children={children}
          />
        </Grid>

        <Grid
          col
          xs="auto"
          alignItems="center"
        >
          <ArrowFlatRight
            className={classSummaryIcon}
            color="success"
          />
        </Grid>
      </Grid>
    </Factory>
  );
});

AccordionSummary.displayName = 'Accordion.Summary';

AccordionSummary.propTypes = {
  /**
   * Add DOM element. Can be a React component or HTML element.
   */
  children: PropTypes.node.isRequired,
  /**
   * Add a new CSS class to `className` property.
   * @default undefined
   */
  className: PropTypes.string,
  /**
   * @borrows Color.propTypes.color as Typography.propTypes.color
   * @default primary
   */
  color: Typography.propTypes.color,
};

/**
 * Add @property {object} defaultProps made available properties information
 * for Props in the Storybook but do not use as major define for default properties.
 */
AccordionSummary.defaultProps = {
  color: 'primary',
};

export default AccordionSummary;
