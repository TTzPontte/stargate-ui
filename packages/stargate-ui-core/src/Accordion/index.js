export { default as AccordionDetails } from './AccordionDetails';
export { default as AccordionSummary } from './AccordionSummary';
export { default } from './Accordion';
