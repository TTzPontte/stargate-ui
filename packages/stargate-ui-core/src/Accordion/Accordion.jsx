import React from 'react';
import PropTypes from 'prop-types';

import AccordionSummary from './AccordionSummary';
import AccordionDetails from './AccordionDetails';
import AccordionContext from './AccordionContext';
import Card from '../Card';

const Accordion = React.forwardRef((props, ref) => {
  const {
    children: defaultChildren,
    onClick,
    controller,
    opened: defaultOpened = false,
    ...cardProps
  } = props;

  const children = React.Children.toArray(defaultChildren);
  const [opened, setOpened] = React.useState(defaultOpened);

  if (controller) {
    React.useEffect(() => {
      setOpened(defaultOpened);
    }, [defaultOpened]);
  }

  const handleToggle = React.useCallback(
    (event) => {
      if (!controller) {
        setOpened(!opened);
      }

      if (onClick) {
        onClick(event, [
          controller,
          controller ? defaultOpened : opened,
        ]);
      }
    },
    [defaultOpened, opened],
  );

  const context = React.useMemo(() => ({
    opened,
    onToggle: handleToggle,
  }), [
    opened,
    onClick,
  ]);

  return (
    <Card
      {...cardProps}
      marginBottom={2}
      paddingX={0}
      paddingY={0}
      borderColor="primary"
    >
      <AccordionContext.Provider
        value={context}
        children={children}
      />
    </Card>
  );
});

Accordion.Summary = AccordionSummary;
Accordion.Details = AccordionDetails;

Accordion.displayName = 'Accordion';

Accordion.propTypes = {
  /**
   * Add component dependencies.
   */
  children: PropTypes.oneOfType([
    PropTypes.instanceOf(Accordion.Summary),
    PropTypes.instanceOf(Accordion.Details),
  ]).isRequired,
  /**
   * Trigger when `Accordion.Summary` is clicked.
   *
   * @param {object} event Default click event
   * @param {array} controller Name of clicked accordion and its state
   *
   * @default undefined
   */
  onClick: PropTypes.func,
  /**
   * Add controller name and if `onClick` is defined, is returned it name
   * to control accordion individually.
   * @default undefined
   */
  controller: PropTypes.string,
  /**
   * Add state of accordion.
   * @default false
   */
  opened: PropTypes.bool,
};

/**
 * Add @property {object} defaultProps made available properties information
 * for Props in the Storybook but do not use as major define for default properties.
 */
Accordion.defaultProps = {
  opened: false,
};

export default Accordion;
