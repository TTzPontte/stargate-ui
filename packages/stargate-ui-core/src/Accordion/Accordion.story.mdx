import {
  Meta,
  Canvas,
  Status,
  Props,
  Story,
  withAction,
} from '@pontte/stargate-ui-playground';

import Accordion from './Accordion';
import AccordionSummary from './AccordionSummary';
import AccordionDetails from './AccordionDetails';
import Typography from '../Typography';

<Meta title="Core/Accordion" component={Accordion} />

# Accordion

<Status type="stable" />

Accordion is a expansible container that may stand alone as a card or piled up.

## Basic Usage

<Canvas>
  <Story name="Basic Usage">
    <Typography>Default variant</Typography>
    <Accordion>
      <AccordionSummary>
        #1 Accordion summary
      </AccordionSummary>
      <AccordionDetails>
        <Typography gutter={0}>
          Lorem, ipsum dolor sit amet consectetur adipisicing elit. Asperiores voluptates laborum assumenda excepturi cupiditate, dicta repellat odio minima expedita unde magni, omnis quaerat animi exercitationem ratione. Velit iure accusantium maxime?
        </Typography>
      </AccordionDetails>
    </Accordion>
    <Accordion>
      <AccordionSummary>
        #2 Accordion summary
      </AccordionSummary>
      <AccordionDetails>
        <Typography gutter={0}>
          Lorem, ipsum dolor sit amet consectetur adipisicing elit. Asperiores voluptates laborum assumenda excepturi cupiditate, dicta repellat odio minima expedita unde magni, omnis quaerat animi exercitationem ratione. Velit iure accusantium maxime?
        </Typography>
      </AccordionDetails>
    </Accordion>
    <Accordion marginBottom={4}>
      <AccordionSummary>
        #3 Accordion summary
      </AccordionSummary>
      <AccordionDetails>
        <Typography gutter={0}>
          Lorem, ipsum dolor sit amet consectetur adipisicing elit. Asperiores voluptates laborum assumenda excepturi cupiditate, dicta repellat odio minima expedita unde magni, omnis quaerat animi exercitationem ratione. Velit iure accusantium maxime?
        </Typography>
      </AccordionDetails>
    </Accordion>
    <Typography>Outlined variant</Typography>
    <Accordion color="primary" variant="outlined">
      <AccordionSummary>
        #1 Accordion summary
      </AccordionSummary>
      <AccordionDetails>
        <Typography gutter={0}>
          Lorem, ipsum dolor sit amet consectetur adipisicing elit. Asperiores voluptates laborum assumenda excepturi cupiditate, dicta repellat odio minima expedita unde magni, omnis quaerat animi exercitationem ratione. Velit iure accusantium maxime?
        </Typography>
      </AccordionDetails>
    </Accordion>
    <Accordion color="primary" variant="outlined">
      <AccordionSummary>
        #2 Accordion summary
      </AccordionSummary>
      <AccordionDetails>
        <Typography gutter={0}>
          Lorem, ipsum dolor sit amet consectetur adipisicing elit. Asperiores voluptates laborum assumenda excepturi cupiditate, dicta repellat odio minima expedita unde magni, omnis quaerat animi exercitationem ratione. Velit iure accusantium maxime?
        </Typography>
      </AccordionDetails>
    </Accordion>
    <Accordion color="primary" variant="outlined">
      <AccordionSummary>
        #3 Accordion summary
      </AccordionSummary>
      <AccordionDetails>
        <Typography gutter={0}>
          Lorem, ipsum dolor sit amet consectetur adipisicing elit. Asperiores voluptates laborum assumenda excepturi cupiditate, dicta repellat odio minima expedita unde magni, omnis quaerat animi exercitationem ratione. Velit iure accusantium maxime?
        </Typography>
      </AccordionDetails>
    </Accordion>
  </Story>
</Canvas>

<Props component={Accordion} />

### Accordion Summary

<Props component={AccordionSummary} />

### Accordion Details

<Props component={AccordionDetails} />

## Controlled accordion

Accordion allows controlling opening individually.

<Canvas>
  <Story name="Controlled accordion">
    {
      (props) => {
        const [controller, setController] = React.useState({ opened: false });
        const onClick = (event, [controller, opened]) => {
          setController({ name: controller, opened: !opened });
        };
        return (
          <React.Fragment>
            <Accordion
              controller="accordion-1"
              opened={controller.name === 'accordion-1' && controller.opened}
              onClick={onClick}
            >
              <AccordionSummary>
                #1 Accordion summary
              </AccordionSummary>
              <AccordionDetails>
                <Typography>
                  Lorem, ipsum dolor sit amet consectetur adipisicing elit. Asperiores voluptates laborum assumenda excepturi cupiditate, dicta repellat odio minima expedita unde magni, omnis quaerat animi exercitationem ratione. Velit iure accusantium maxime?
                </Typography>
              </AccordionDetails>
            </Accordion>
            <Accordion
              controller="accordion-2"
              opened={controller.name === 'accordion-2' && controller.opened}
              onClick={onClick}
            >
              <AccordionSummary>
                #2 Accordion summary
              </AccordionSummary>
              <AccordionDetails>
                <Typography>
                  Lorem, ipsum dolor sit amet consectetur adipisicing elit. Asperiores voluptates laborum assumenda excepturi cupiditate, dicta repellat odio minima expedita unde magni, omnis quaerat animi exercitationem ratione. Velit iure accusantium maxime?
                </Typography>
              </AccordionDetails>
            </Accordion>
            <Accordion
              controller="accordion-3"
              opened={controller.name === 'accordion-3' && controller.opened}
              onClick={onClick}
            >
              <AccordionSummary>
                #3 Accordion summary
              </AccordionSummary>
              <AccordionDetails>
                <Typography>
                  Lorem, ipsum dolor sit amet consectetur adipisicing elit. Asperiores voluptates laborum assumenda excepturi cupiditate, dicta repellat odio minima expedita unde magni, omnis quaerat animi exercitationem ratione. Velit iure accusantium maxime?
                </Typography>
              </AccordionDetails>
            </Accordion>
          </React.Fragment>
        )
      }
    }
  </Story>
</Canvas>
