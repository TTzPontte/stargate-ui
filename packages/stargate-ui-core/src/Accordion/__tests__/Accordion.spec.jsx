import React from 'react';

import Accordion from '../Accordion';
import AccordionSummary from '../AccordionSummary';
import AccordionDetails from '../AccordionDetails';

describe('Accordion', () => {
  it('should render properly', () => {
    const { container } = render(
      <Accordion>
        <AccordionSummary>
          Accordion Summary
        </AccordionSummary>
        <AccordionDetails>
          Lorem ipsum dolor sit amet, consectetur adipisicing elit. Aut voluptate ad distinctio quas vel praesentium, est vero suscipit dignissimos, reprehenderit, cumque nam incidunt reiciendis quibusdam nihil dolorem perferendis eos accusamus?
        </AccordionDetails>
      </Accordion>,
    );

    expect(container).toMatchSnapshot();
  });
});
