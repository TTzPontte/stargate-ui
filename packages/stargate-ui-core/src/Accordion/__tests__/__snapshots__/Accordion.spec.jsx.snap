// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should render properly 1`] = `
<div>
  <div
    class="stargate-ui-factory-0-2-4 stargate-ui-factory-d0-0-2-6 stargate-ui-accordionVariantDefault-0-2-1"
  >
    <button
      class="stargate-ui-factory-0-2-13 stargate-ui-factory-d0-0-2-15 stargate-ui-summary-0-2-8 stargate-ui-summaryVariantDefault-0-2-11"
      type="button"
    >
      <div
        class="stargate-ui-factory-0-2-35 stargate-ui-factory-d0-0-2-37 stargate-ui-gridRow-0-2-17 stargate-ui-gridRow-d0-0-2-26 stargate-ui-gridRowGutter-0-2-19 stargate-ui-gridRowGutter-d2-0-2-28"
      >
        <div
          class="stargate-ui-factory-0-2-57 stargate-ui-factory-d0-0-2-59 stargate-ui-gridCol-0-2-40 stargate-ui-gridCol-d1-0-2-49 stargate-ui-gridColSize-xs-0-2-42 stargate-ui-gridColSize-xs-d3-0-2-51"
        >
          <p
            class="stargate-ui-factory-0-2-68 stargate-ui-factory-d0-0-2-70 stargate-ui-typography-0-2-61 stargate-ui-typography-d0-0-2-65 stargate-ui-typographyFluid-0-2-64 stargate-ui-typographyFluid-d2-0-2-67"
          >
            Accordion Summary
          </p>
        </div>
        <div
          class="stargate-ui-factory-0-2-90 stargate-ui-factory-d0-0-2-92 stargate-ui-gridCol-0-2-73 stargate-ui-gridCol-d1-0-2-82 stargate-ui-gridColSize-xs-0-2-75 stargate-ui-gridColSize-xs-d3-0-2-84"
        >
          <span
            aria-hidden="true"
            class="stargate-ui-factory-0-2-97 stargate-ui-factory-d0-0-2-99 stargate-ui-svg-0-2-96 stargate-ui-svgIcon-0-2-94 stargate-ui-svgIcon-d0-0-2-95 stargate-ui-summaryIcon-0-2-9"
          >
            <svg
              class="stargate-ui-factory-0-2-101 stargate-ui-factory-d0-0-2-103"
              viewBox="0,0,24,24"
            >
              <path
                d="M12 18.35A4.19 4.19 0 019.09 17L.34 7.83a1.24 1.24 0 111.8-1.71l8.76 9.22a1.67 1.67 0 001.11.53 1.54 1.54 0 00.92-.51L21.86 6a1.24 1.24 0 111.8 1.72l-8.91 9.3A4 4 0 0112 18.35z"
              />
            </svg>
          </span>
        </div>
      </div>
    </button>
    <div
      class="stargate-ui-factory-0-2-110 stargate-ui-factory-d0-0-2-112 stargate-ui-details-0-2-105 stargate-ui-detailsVariantDefault-0-2-109"
    >
      <div
        class="stargate-ui-detailsContent-0-2-106"
      >
        Lorem ipsum dolor sit amet, consectetur adipisicing elit. Aut voluptate ad distinctio quas vel praesentium, est vero suscipit dignissimos, reprehenderit, cumque nam incidunt reiciendis quibusdam nihil dolorem perferendis eos accusamus?
      </div>
    </div>
  </div>
</div>
`;
